src/smt/theory_str_noodler/decision_procedure.cpp

```diff
@@ -784,40 +789,57 @@ namespace smt::noodler {
+                STRACE("str-conversion-int", tout << "contains-non-digit NFA:" << std::endl << *aut_non_valid_part << std::endl;);
+                if (!aut_non_valid_part->is_lang_empty()) {
+                    // aut_non_valid_part is language of words that contain at least one non-digit
+                    //  - we can get here only for the case that conv.type is to_int (for from_int, by assumptions, s should be restricted to language of "valid numbers + empty string")
+                    //  - if s_i = one of these words, then s must also contain non-digit => result i = -1
+                    // we therefore create following formula:
+                    //       |s_i| is length of some word from aut_non_valid_part && int_version_of(s_i) = -1 && i = -1
+                    result.succ.emplace_back(LenFormulaType::AND, std::vector<LenNode>{ solution.aut_ass.get_lengths(*aut_non_valid_part, subst_var), LenNode(LenFormulaType::EQ, { int_version_of(subst_var), -1 }), LenNode(LenFormulaType::EQ, {i, -1}) });
+                    if (code_subst_vars.contains(subst_var)) {
+                        // s_i is used in some to_code/from_code
+                        // => we need to add to the previous formula also the fact, that s_i cannot encode code point of a digit
+                        //      .. && !(48 <= code_version_of(s_i) <= 57)
+                        result.succ.back().succ.emplace_back(LenFormulaType::LT, std::vector<LenNode>{ code_version_of(subst_var), 48 });
+                        result.succ.back().succ.emplace_back(LenFormulaType::LT, std::vector<LenNode>{ 57, code_version_of(subst_var) });
+                    }
                 }
+            } else {
+                // for from_int, we assume that s is restricted to language of "valid numbers + empty string", which means that
+                // s_i should also be restricted to this language => 'aut intersected with only_digits == aut'
+                aut_valid_part = aut;
+                STRACE("str-conversion-int", tout << "only-digit NFA:" << std::endl << *aut_valid_part << std::endl;);
             }
 
+            unsigned max_length_of_words = aut_valid_part->num_of_states();
+
             // we want to enumerate all words containing digits -> cannot be infinite language
-            if (!aut_valid_part.is_acyclic()) {
-                STRACE("str-conversion", tout << "failing NFA:" << std::endl << aut_valid_part << std::endl;);
-                util::throw_error("cannot process to_int/from_int for automaton with infinite language");
+            if (!aut_valid_part->is_acyclic()) {
+                STRACE("str-conversion", tout << "failing NFA:" << std::endl << *aut_valid_part << std::endl;);
+                // util::throw_error("cannot process to_int/from_int for automaton with infinite language");
+                res_precision = LenNodePrecision::UNDERAPPROX;
+                if (max_length_of_words > 3) {
```

Comment:
As a future work, we could propagate the LIA solver here and check, if the formula `subst_var < 100` is valid. In that case we have a precise solution and we could set `res_precision` to `PRECISE`.

Reply:
`subst_var` is a string var, so we would have to check something else, probably you meant `|subst_var| < 3`.
But I am not sure if it would work.
I was thinking about doing something a bit similar, returning multiple formulae (one underapproximation, one something like overapproximation), but it is quite complicated and it is not that clear how it should work.

Anyway, for `stringfuzz`, the best thing would be to transform `n = str.to_int(x)` on the input to `n = str.to_int(x) & x ∈ 0*n` (and also do that for lenghts, i.e. `n = str.len(x) & x ∈ .{n}`), I think that would solve everything.
So doing something with overapproximation or something similar would only be helpful for `full_str_int`, and it is questionable how helpful.

Reply:
1) ok, and what about the same checking for `from_int`? Is it possible at all in our setting? The replacement of `from_int` by `to_int` is done in relevant_eh, right ?
2) Agree, there is a good hope that this can be done in seq rewriter.

Reply:
1. I do not understand the question. What do you mean by same checking? Also `from_int` is not replaced with `to_int`, I do handle them in similar way during the LIA formula creation, but it is still important to know if I have `to_int` or `from_int` (for handling invalid cases).

Reply:
Sorry, I wrote it confusedly. I though that the current problem (mainly for `full_str_int`) are constraints of the form `w = from_int(x)` where `x` is some integer variable. Because we do not know about the possible values of `x`, we are not currently able to solve it, right? But if we find out that `x` is always smaller than 100, we can enumerate it quite easily.

Reply:
Hmm, I think I understand now.  It can probably be somehow applied for it, I'll have to think about it.

Anyway, there are only 363 formulae with `from_int` in `full_str_int`, so I don't think it's the main problem for this benchmark right now.
You can see that the underapproximation, which is useful only for SAT cases (see how number of UNSAT cases did not change at all), already helped a lot with `unknown`s, however, most of them moved to TO/MO, so I think we also need to focus on simplifying the formula somehow.